### PR TITLE
Prevent overwriting existing host_uuid file

### PR DESCRIPTION
**Pull request by rosstimothy**
In some circumstances, multiple Teleport processes may be trying to write the host_uuid file in the same data directory simultaneously. The last of the writers would win, and any process using a host UUID that did not match what ended up on disk could get into a perpetual state of being unable to connect to the cluster.

To avoid the raciness, the host_uuid file writing process is no longer a blind upsert. Instead, special care is taken to ensure that there can only be a single writer, and that any subsequent updates to the file are aborted and the first value written is used instead.

changelog: Prevent two Teleport processes from overwriting the same host_uuid file.

### CODE REVIEW

File: lib/utils/utils.go [hunk 1]
```diff
@@ -512,29 +513,75 @@ func WriteHostUUID(dataDir string, id string) error {
...
+
+		id = rawID.String()
+
+		writeFile := func() (string, error) {
+			f, err := os.OpenFile(hostUUIDLock, os.O_WRONLY|os.O_CREATE, 0o400)
+			if err != nil {
+				return "", trace.Wrap(err)
+			}
+
+			unlock, err := FSTryWriteLock(f.Name())
```
Comment (espadolini) [inline lib/utils/utils.go:550]: I believe that `FSTryWriteLock` will create (and clean up?) the lock file on its own, so there's no need to create it or delete it here - moreover, we should not delete the lockfile if there's been errors, I think?

File: lib/utils/utils.go [hunk 2]
```diff
@@ -512,29 +513,75 @@ func WriteHostUUID(dataDir string, id string) error {
...
+			}
+
+			defer func() {
+				os.Remove(hostUUIDLock)
+				unlock()
+			}()
+
+			if read, err := ReadHostUUID(dataDir); err == nil {
+				return read, nil
+			}
```
Comment (espadolini) [inline lib/utils/utils.go:562]:
```suggestion
			if read, err := ReadHostUUID(dataDir); err == nil {
				return read, nil
			} else if !trace.IsNotExists(err) {
				return "", trace.Wrap(err)
			}
```

File: lib/utils/hostid/hostid.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+			if err := renameio.WriteFile(hostUUIDFileName, []byte(id), 0o400); err != nil {
+				return "", trace.Wrap(err)
+			}
+
+			return id, nil
+		}
+
+		id, err = writeFile()
+		if err != nil {
+			if errors.Is(err, fs.ErrPermission) || errors.Is(err, utils.ErrUnsuccessfulLockTry) {
```
Comment (espadolini) [inline lib/utils/hostid/hostid.go:132]: What's the rationale for retrying on a permission error?

File: lib/utils/hostid/hostid.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+		if err != nil {
+			return "", trace.BadParameter("" +
+				"Teleport failed to generate host UUID. " +
+				"This may happen if randomness source is not fully initialized when the node is starting up. " +
+				"Please try restarting Teleport again.")
+		}
+
+		id = rawID.String()
+
+		writeFile := func() (string, error) {
```
Comment (espadolini) [inline lib/utils/hostid/hostid.go:110]: Something like this? And then we can do `id, err := writeFile(rawID.String())` or something along those lines.

```suggestion
		writeFile := func(potentialNewID string) (actualID string, error) {
```

File: lib/utils/hostid/hostid.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+	id := strings.TrimSpace(string(out))
+	if id == "" {
+		return "", trace.NotFound("host uuid is empty")
+	}
+	return id, nil
+}
+
+// WriteFile writes host UUID into a file
+func WriteFile(dataDir string, id string) error {
+	err := os.WriteFile(GetPath(dataDir), []byte(id), os.ModeExclusive|0400)
```
Comment (espadolini) [inline lib/utils/hostid/hostid.go:69]:
```suggestion
	err := renameio.WriteFile(GetPath(dataDir), []byte(id), 0o400)
```

**Review by espadolini — APPROVED** [review]

**Review by greedy52 — APPROVED** [review]

**Pull request merged**